packages/connect/src/utils/debug.ts

```diff
@@ -14,63 +14,116 @@ const colors: Record<string, string> = {
     '@trezor/transport': 'color: #bada55; background: #000;',
 };
 
-type LogMessage = {
+export type LogMessage = {
     level: string;
     prefix: string;
     message: any[];
     timestamp: number;
+    css: string;
+};
+
+export type LogWriter = {
+    add: (message: LogMessage) => void;
 };
 
 const MAX_ENTRIES = 100;
 
+const stringify = (obj: Record<string, any>) => {
+    let cache: string[] = [];
+    const str = JSON.stringify(obj, (_key, value) => {
+        if (typeof value === 'object' && value !== null) {
+            if (cache.indexOf(value) !== -1) {
+                // Circular reference found, discard key
+                return;
+            }
+            // Store value in our collection
+            cache.push(value);
+        }
+        return value;
+    });
+    cache = [];
+    return str;
+};
+
```

Comment:
lets keep discussion about this open. I don't like stringifying our super big interconnected objects. we should come up with something better

Reply:
Sure, we can keep discussing what is the best approach for this.

Reply:
So using `device.toMessageObject()` in DeviceList apparently removes the issue of circular dependencies and no stringify is required at all.

https://github.com/trezor/trezor-suite/pull/8623/commits/f123e29f312e82cc94d3b3916c8a0fbb76efc10e

Reply:
ok, just lets not forget to have some mechanism for quickly catching such problems ever in the future.